docs(metrics-operator): usage of SLI and SLO converters (#2013)
keptn/lifecycle-toolkit: metrics-operator/converter/slo_converter.md

# SLO Converter

## Description

SLO converter is a tool to convert `slo.yaml` files used in [KeptnV1](https://v1.keptn.sh/) into the new
`AnalysisDefinition` resources used in the new kubernetes-native [Keptn](https://lifecycle.keptn.sh/).
The converter is part of `metrics-operator` image.

## Usage

The converter will convert a single `slo.yaml` file into single `AnalysisDefintion` resource.

To run the converter, execute the following command:

> ```shell

Reviewer: same suggestion as above, let's define the placeholders as env vars

docker-run <METRICS_OPERATOR_IMAGE> manager --convert-slo=<PATH_TO_SLO> --slo-namespace=<ANALYSIS_VALUE_TEMPLATE_NAMESPACE> --definition=<DEFINITION_NAME>
```

Please be aware, you need to substitute the placeholders with the following information:

* **PATH_TO_SLO** - path to your `slo.yaml` file
* **ANALYSIS_VALUE_TEMPLATE_NAMESPACE** - namespace of `AnalysisValueTemplate` which will be referenced in objectives
* **DEFINITION_NAME** - name of created `AnalysisDefinition`

> **Note**

All the SLOs present in `slo.yaml` file will reference `AnalysisValueTemplate` resources from the namespace defined

> by `ANALYSIS_VALUE_TEMPLATE_NAMESPACE` argument.

Reviewer: will the AnalysisDefinitions then always be created in the same namespace as the AnalysisValueTemplates?

Author: the analysisdefinition does not have it's namespace defined, the namespace is defined only for the referenced `AnalysisValueTemplates`. The `AnalysisDefinition` itself can be applied to any namespace

## Conversion details

We have multiple use-cases which are and which are not supported.
There is a need to convert the use-cases that make
logical sense and are common, but in some cases, where it is problematic and these cases will not be supported.

> **Note** Please be aware, that comparison criteria containing `%` symbol ware not supported and will be ignored.

### Unsupported use-cases

Criteria with 3 and more inputs won't be supported, only the first 2 non-percentage inputs (those not containing `%`
as we do not support comparison rules) will be taken and converted.
In the example below, only `<600` and `>400` rules will be converted.
Rule `>800` will be ignored.

```yaml

> objectives:

Reviewer: let's also add examples for how the resulting AnalysisDefinition for SLOs containing unsupported criteria would look like

Reviewer: edit: I just saw that there is a example of a complete SLO -> AnalysisDefinition is available below. But I think it might also help to see the examples for the individual examples here to make it easier to see how such objectives end up looking like after the conversion

- sli: response_time_p95
  displayName: "Response Time P95"
  pass:
  - criteria
    - "<600"
    - ">400"
    - ">800"
  weight: 2
  key_sli: true
```

Secondly, conversion of multiple criteria elements won't be supported at all.
These criteria are combined with logical OR operator (see documentation [here](https://github.com/keptn/spec/blob/master/service_level_objective.md#objectives)).

> Only a single criteria element rules will be supported.

Reviewer: ```suggestion
Only single criteria element rules will be supported.
```

> These criteria will be converted into informative objectives.

Reviewer: This might be a bit confusing - does "these" refer to the single criteria element rules mentioned in the sentence before that, or to the multiple criteria elements? I guess the latter, but we should be clearer here. Maybe something like:

Second, only rules containing a single criteria element will be supported. Rules containing multiple criteria elements can not be converted and will be turned into informative items.
Informative in this context means that there are no objectives that have to be met and therefore the item will not affect the overall score. However, their values are still retrieved to provide additional insights for the execution of an Analysis.

> These criteria will be converted into informative objectives.

Reviewer: "Informative objectives" sounds a bit like a contradiction. Also, we should clarify what "informative" means in that context. In my other comment I have tried to elaborate a bit on that: https://github.com/keptn/lifecycle-toolkit/pull/2013/files#r1314879206

> An example of unsupported SLOs for conversion here:

Reviewer: ```suggestion
An example of an unsupported SLOs for conversion is the following:
```

# SLO Converter

> ## Description

Reviewer 2: Instead of a description I'd rather add a TOC since the file is quite long

Reviewer 2: you can use https://www.npmjs.com/package/markdown-toc

> SLO converter is a tool to convert `slo.yaml` files used in [KeptnV1](https://v1.keptn.sh/) into the new
> `AnalysisDefinition` resources used in the new kubernetes-native [Keptn](https://lifecycle.keptn.sh/).

Reviewer 2: ```suggestion
SLO converter is a tool to convert `slo.yaml` files used in [KeptnV1](https://v1.keptn.sh/) into the new
`AnalysisDefinition` resources used in the new kubernetes-native [Keptn](https://lifecycle.keptn.sh/).
```

```suggestion
SLO converter is a tool to convert the old `slo.yaml` [file format](https://github.com/keptn/spec/blob/master/service_level_objective.md) into the new
`AnalysisDefinition ` [custom resource definition](../../docs/content/en/docs/crd-ref/metrics/v1alpha3/_index.md).
```

* **PATH_TO_SLO** - path to your `slo.yaml` file
* **ANALYSIS_VALUE_TEMPLATE_NAMESPACE** - namespace of `AnalysisValueTemplate` which will be referenced in objectives
* **DEFINITION_NAME** - name of created `AnalysisDefinition`

> **Note**

All the SLOs present in `slo.yaml` file will reference `AnalysisValueTemplate` resources from the namespace defined
by `ANALYSIS_VALUE_TEMPLATE_NAMESPACE` argument.

## Conversion details

We have multiple use-cases which are and which are not supported.
There is a need to convert the use-cases that make
logical sense and are common, but in some cases, where it is problematic and these cases will not be supported.

> **Note** Please be aware, that comparison criteria containing `%` symbol ware not supported and will be ignored.

### Unsupported use-cases

Criteria with 3 and more inputs won't be supported, only the first 2 non-percentage inputs (those not containing `%`
as we do not support comparison rules) will be taken and converted.
In the example below, only `<600` and `>400` rules will be converted.
Rule `>800` will be ignored.

```yaml
objectives:
- sli: response_time_p95
  displayName: "Response Time P95"
  pass:

>   - criteria

Reviewer 2: ```suggestion
  - criteria:
```

    - "<600"
    - ">400"
    - ">800"
  weight: 2
  key_sli: true
```

Secondly, conversion of multiple criteria elements won't be supported at all.
Rules containing multiple criteria elements combined with logical OR (see documentation
[here](https://github.com/keptn/spec/blob/master/service_level_objective.md#objectives))

> can not be converted and will be turned into informative items.

Reviewer 2: why can NOT we convert it?! I would expect that since there are multiple criteria, we will have multiple objectives 🤔

Author: You have a point, some combinations with OR operator can be converted, but only those, where we can use inRange/notInRange operators and they cannot be splitted into multiple objectives, that won't work good

Author: partially covered with https://github.com/keptn/lifecycle-toolkit/issues/2021

Informative in this context means that there are no objectives that have to be met and
therefore the item will not affect the overall score.
However, their values are still retrieved to provide additional insights for the execution of an Analysis.
An example of an unsupported SLOs for conversion is the following:

```yaml
objectives:
- sli: response_time_p95
  displayName: "Response Time P95"
  pass:
  - criteria:
    - "<=+10%"

>   - criteria

Reviewer 2: ```suggestion
  - criteria:
```

    - "<600"
  weight: 2
  key_sli: true
```

> Thirdly, a case where `pass` criteria are set, `warn` criteria are set, but pass criteria interval and warn

Reviewer 2: instead of secondly, thirdly, I'd would suggest having a proper name for these cases and add a subsection for them. This way, it is easier with a TOC on top of the document to jump directly to the conversion case

```yaml
objectives:
- sli: response_time_p95
  displayName: "Response Time P95"
  pass:
  - criteria:
    - ">400"
    - "<600"
  warn:
  - criteria:
    - ">200"
    - "<800" 
```

will be converted to

```yaml
spec:
  objectives:
  - analysisValueTemplateRef:
      name: response_time_p95
      namespace: default
    target:
      failure:
        notInRange:
          highBound: "200"
          lowBound: "800"
      warning:
        notInRange:
          highBound: "400"
          lowBound: "600"
```

1. `pass` criteria set, `warn` criteria set, pass criteria interval is superset of warn criteria interval

```yaml
objectives:
- sli: response_time_p95
  displayName: "Response Time P95"
  pass:
  - criteria:
    - ">200"
    - "<800"
  warn:
  - criteria:
    - ">400"
    - "<600" 
```

will be converted to

```yaml
spec:
  objectives:
  - analysisValueTemplateRef:
      name: response_time_p95
      namespace: default
    target:
      failure:
        notInRange:
          highBound: "200"
          lowBound: "800"
      warning:
        inRange:
          highBound: "400"
          lowBound: "600"
```

> ## Example

Reviewer 2: I will move this example before the limitation

Author: I would keep it in the end, we have partial examples with the limitations and it's quite long and complex

  include_result_with_score: "pass"
  number_of_comparison_results: 1
objectives:
  - sli: "response_time_p90"
    key_sli: false
    pass:
    - criteria:
        - ">600"
        - "<800"
    warning:
    - criteria:
        - "<=1000"
        - ">500"
    weight: 2
  - sli: "response_time_p80"
    key_sli: false
    pass:
      - criteria:
          - ">600"
          - "<800"
    warning:
      - criteria:
          - "<=1000"
    weight: 2
  - sli: "response_time_p70"
    key_sli: false
    warning:
      - criteria:
          - ">600"
          - "<800"
    pass:
      - criteria:
          - "<=1000"
    weight: 2
  - sli: "response_time_p95"
    key_sli: false
    pass:
      - criteria:
          - "<=+75%"
          - "<800"
    warning:
      - criteria:
          - "<=1000"
          - "<=+100%"
    weight: 1
  - sli: "cpu"
    pass:
      - criteria:
          - "<=+100%"
          - ">=80"
      - criteria:
          - "<=+100%"
          - ">=80"
  - sli: "throughput"
    pass:
      - criteria:
          - "<=+100%"
          - ">=-80%"
  - sli: "error_rate"
total_score:
  pass: "100%"
  warning: "65%"
```

with the following command

```shell

> docker-run $METRICS_OPERATOR_IMAGE manager --convert-slo=./slo.yaml --slo-namespace=default --definition=defName

Reviewer 2: similarly to before, please use the ghcr registry with `latest` as tag here

  - [Table of Contents](#table-of-contents)
  - [Description](#description)
  - [Usage](#usage)
  - [Conversion details](#conversion-details)
    - [Unsupported use-cases](#unsupported-use-cases)
      - [Criteria with 3 and more rules](#criteria-with-3-and-more-rules)
      - [Criteria with intervals without intersection](#criteria-with-intervals-without-intersection)
    - [Supported use-cases](#supported-use-cases)
      - [Single rule criteria](#single-rule-criteria)
      - [Criteria combined with logical OR operator](#criteria-combined-with-logical-or-operator)
      - [Criteria with intervals with intersection](#criteria-with-intervals-with-intersection)
  - [Example](#example)

## Description

SLO converter is a tool to convert the old `slo.yaml`
[file format](https://github.com/keptn/spec/blob/master/service_level_objective.md) into the new
`AnalysisDefinition` [custom resource definition](../../docs/content/en/docs/crd-ref/metrics/v1alpha3/_index.md).
The converter is part of `metrics-operator` image.

## Usage

The converter will convert a single `slo.yaml` file into single `AnalysisDefintion` resource.

To run the converter, execute the following command:

```shell

> METRICS_OPERATOR_IMAGE=ghcr.io/keptn/lifecycle-operator:latest

Reviewer 3: same here

# SLO Converter

> ## Table of Contents
> 
> - [SLO Converter](#slo-converter)
>   - [Table of Contents](#table-of-contents)
>   - [Description](#description)
>   - [Usage](#usage)
>   - [Conversion details](#conversion-details)
>     - [Unsupported use-cases](#unsupported-use-cases)
>       - [Criteria with 3 and more rules](#criteria-with-3-and-more-rules)
>       - [Criteria with intervals without intersection](#criteria-with-intervals-without-intersection)
>     - [Supported use-cases](#supported-use-cases)
>       - [Single rule criteria](#single-rule-criteria)
>       - [Criteria combined with logical OR operator](#criteria-combined-with-logical-or-operator)
>       - [Criteria with intervals with intersection](#criteria-with-intervals-with-intersection)
>   - [Example](#example)

Reviewer 2: to make it easier using tools to keep this up2date, I'd prefer having the `<!-- toc -->` comments.

Also, we can compress it 

```suggestion
<details>
<summary>Table of Contents</summary>

<!-- toc -->

- [SLO Converter](#slo-converter)
  - [Table of Contents](#table-of-contents)
  - [Description](#description)
  - [Usage](#usage)
  - [Conversion details](#conversion-details)
    - [Unsupported use-cases](#unsupported-use-cases)
      - [Criteria with 3 and more rules](#criteria-with-3-and-more-rules)
      - [Criteria with intervals without intersection](#criteria-with-intervals-without-intersection)
    - [Supported use-cases](#supported-use-cases)
      - [Single rule criteria](#single-rule-criteria)
      - [Criteria combined with logical OR operator](#criteria-combined-with-logical-or-operator)
      - [Criteria with intervals with intersection](#criteria-with-intervals-with-intersection)
  - [Example](#example)
  
<!-- tocstop -->

</details>
```

## Description

SLO converter is a tool to convert the old `slo.yaml`
[file format](https://github.com/keptn/spec/blob/master/service_level_objective.md) into the new
`AnalysisDefinition` [custom resource definition](../../docs/content/en/docs/crd-ref/metrics/v1alpha3/_index.md).
The converter is part of `metrics-operator` image.

## Usage

The converter will convert a single `slo.yaml` file into single `AnalysisDefintion` resource.

To run the converter, execute the following command:

<!---x-release-please-start-version-->
```shell
METRICS_OPERATOR_IMAGE=ghcr.io/keptn/metrics-operator:v0.8.2
PATH_TO_SLO=<PATH_TO_SLO>
ANALYSIS_VALUE_TEMPLATE_NAMESPACE=<ANALYSIS_VALUE_TEMPLATE_NAMESPACE>
ANALYSIS_DEFINITION_NAME=<ANALYSIS_DEFINITION_NAME>

> docker-run $METRICS_OPERATOR_IMAGE manager --convert-slo=$PATH_TO_SLO --analysis-value-template-namespace=$ANALYSIS_VALUE_TEMPLATE_NAMESPACE --analysis-definition-name=$ANALYSIS_DEFINITION_NAME

Reviewer 2: ```suggestion
docker run $METRICS_OPERATOR_IMAGE manager --convert-slo=$PATH_TO_SLO --analysis-value-template-namespace=$ANALYSIS_VALUE_TEMPLATE_NAMESPACE --analysis-definition-name=$ANALYSIS_DEFINITION_NAME
```

    pass:
    - criteria:
        - ">600"
        - "<800"
    warning:
    - criteria:
        - "<=1000"
        - ">500"
    weight: 2
  - sli: "response_time_p91"
    key_sli: false
    pass:
      - criteria:
        - "<600"
      - criteria:
        - ">800"
    weight: 5
  - sli: "response_time_p80"
    key_sli: false
    pass:
      - criteria:
          - ">600"
          - "<800"
    warning:
      - criteria:
          - "<=1000"
    weight: 2
  - sli: "response_time_p70"
    key_sli: false
    warning:
      - criteria:
          - ">600"
          - "<800"
    pass:
      - criteria:
          - "<=1000"
    weight: 2
  - sli: "response_time_p95"
    key_sli: false
    pass:
      - criteria:
          - "<=+75%"
          - "<800"
    warning:
      - criteria:
          - "<=1000"
          - "<=+100%"
    weight: 1
  - sli: "cpu"
    pass:
      - criteria:
          - "<=+100%"
          - ">=100"
      - criteria:
          - "<=+100%"
          - "<=80"
  - sli: "throughput"
    pass:
      - criteria:
          - "<=+100%"
          - ">=-80%"
  - sli: "error_rate"
total_score:
  pass: "100%"
  warning: "65%"
```

with the following command

```shell

> docker-run $METRICS_OPERATOR_IMAGE manager --convert-slo=./slo.yaml --slo-namespace=default --definition=defName

Reviewer 2: ```suggestion
docker run $METRICS_OPERATOR_IMAGE manager --convert-slo=./slo.yaml --slo-namespace=default --definition=defName
```
